### PR TITLE
UCS/DATASTRUCT: Add allow_list data struct

### PR DESCRIPTION
**Pull request by gleon99**
## What
Introduce the `allow_list` data struct in UCP config parser. It is based on a strings array, adding a `mode`, which is one of `ALL`, `ALLOW` or `NEGATE`.

## Why ?
In order to support negation in `UCX_TLS` and maybe other options.

### CODE REVIEW

File: src/ucs/config/parser.c [hunk 1]
```diff
@@ -855,6 +857,162 @@ void ucs_config_help_array(char *buf, size_t max, const void *arg)
...
+    saveptr = NULL;
+    token   = strtok_r(str_dup + (field->mode == UCS_CONFIG_ALLOW_LIST_NEGATE), ",",
+                       &saveptr);
+
+    i = 0;
+    while (token != NULL) {
+        if (!strcmp(token, ucs_config_parser_all)) {
+            if (field->mode == UCS_CONFIG_ALLOW_LIST_NEGATE || temp_field) {
+                /* Either "negate all" or putting "all" after other items - doesn't make sense */
+                return UCS_ERR_INVALID_PARAM;
```
Comment (hoopoepg) [inline src/ucs/config/parser.c:889]: memory leak?

File: src/ucs/config/parser.c [hunk 2]
```diff
@@ -855,6 +857,162 @@ void ucs_config_help_array(char *buf, size_t max, const void *arg)
...
+            } else {
+                field->mode        = UCS_CONFIG_ALLOW_LIST_ALL;
+                field->array.count = 0;
+                return 1;
```
Comment (hoopoepg) [inline src/ucs/config/parser.c:893]: memory leak?
str_dup is not free

File: src/ucs/config/parser.c [hunk 3]
```diff
@@ -855,6 +857,162 @@ void ucs_config_help_array(char *buf, size_t max, const void *arg)
...
+int ucs_config_sscanf_allow_list(const char *buf, void *dest, const void *arg)
+{
+    ucs_config_allow_list_t *field  = dest;
+    void *temp_field                = NULL;
+    const ucs_config_array_t *array = arg;
+    char *str_dup, *token, *saveptr;
+    int ret;
+    unsigned i;
+
+    str_dup = strdup(buf);
```
Comment (hoopoepg) [inline src/ucs/config/parser.c:869]: ucs_strdup

File: src/ucs/config/parser.c [hunk 4]
```diff
@@ -855,6 +857,162 @@ void ucs_config_help_array(char *buf, size_t max, const void *arg)
...
+    if (allow_list->mode == UCS_CONFIG_ALLOW_LIST_ALL) {
+        snprintf(buf, max, ucs_config_parser_all);
+        return 1;
+    } else if (allow_list->mode == UCS_CONFIG_ALLOW_LIST_NEGATE) {
+        buf[offset++] = ucs_config_parser_negate;
+        max--;
+    }
+
+    for (i = 0; i < allow_list->array.count; ++i) {
+        if (i > 0 && offset < max) {
```
Comment (hoopoepg) [inline src/ucs/config/parser.c:941]: if ((i > 0) && (offset < max))

File: src/ucs/config/parser.c (hunk 3 above)
Comment (gleon99) [reply src/ucs/config/parser.c:869]: I copied this code from `ucs_config_sscanf_array`, which uses `strdup`.
So, if we change only here to `ucs_strdup` it will be inconsistent. Or we change there as well. What do you say?

File: src/ucs/config/types.h [hunk 5]
```diff
@@ -111,6 +111,18 @@ typedef enum {
 
 typedef UCS_CONFIG_STRING_ARRAY_FIELD(names) ucs_config_names_array_t;
 
+typedef enum {
+    UCS_CONFIG_ALLOW_LIST_ALL    = UCS_BIT(0),
+    UCS_CONFIG_ALLOW_LIST_ALLOW  = UCS_BIT(1),
+    UCS_CONFIG_ALLOW_LIST_NEGATE = UCS_BIT(2)
```
Comment (brminich) [inline src/ucs/config/types.h:117]: these options are mutually exclusive. I'd define just one of them, for instance `UCS_CONFIG_ALLOW_LIST_ALLOW`. Then is this flag is not set, it is negate list

File: src/ucs/config/parser.c [hunk 6]
```diff
@@ -855,6 +857,164 @@ void ucs_config_help_array(char *buf, size_t max, const void *arg)
...
+
+void ucs_config_release_allow_list(void *ptr, const void *arg)
+{
+    ucs_config_allow_list_t *allow_list = ptr;
+    const ucs_config_array_t *array     = arg;
+    unsigned i;
+
+    if (allow_list->mode != UCS_CONFIG_ALLOW_LIST_ALLOW) {
+        return;
+    }
```
Comment (brminich) [inline src/ucs/config/parser.c:999]: why no release for negate list?

**Review by brminich — COMMENTED** [review]
general comment: is it possible to unite new parser type handling with the existing `array` type?
(use common inline functions, etc)

File: src/ucs/config/types.h (hunk 5 above)
Comment (gleon99) [reply src/ucs/config/types.h:117]: @brminich ,
From a logical perspective, there are 3 possible, **mutually exclusive** modes:
allow, allow-all, negate.

Obviously 3 bits are not "needed" to represent it and 2 are enough, but it is rather a matter of convenience.
In current implementation we can write:

```
if (allow_list->mode == UCS_CONFIG_ALLOW_LIST_NEGATE) { ...
```

According to what you suggest, it will be harder to check:
```
if (!(allow_list->mode & UCS_CONFIG_ALLOW_LIST_ALLOW) &&
    !(allow_list->mode & UCS_CONFIG_ALLOW_LIST_ALLOW_ALL)) { ...
```
(Or any other logically similar condition - it doesn't matter, the point is that 2 conditions will be needed).
Of course we can "wrap it in macro/function", but even this:
```
if (UCS_CONFIG_ALLOW_LIST_IS_NEGATE_MODE(allow_list->mode)) { ...
```
IMO looks worse than the current variant.

Are you sure that's what we want?

Comment (brminich) [reply src/ucs/config/types.h]: I think it is more consistent with other options in the code.
@yosefe, wdyt?

**Comment by gleon99** [comment]
bot:pipe:retest

**Comment by yosefe** [comment]
smth is off, maybe Java exclusion broke it?
```
2020-12-16T09:42:29.8628129Z Configuration file(s) do(es) not support C++: /__w/1/s/.clang-format
2020-12-16T09:42:29.8647775Z error: `clang-format -lines=28:29 -lines=52:52 -lines=861:1018 -assume-filename=src/ucs/config/parser.c` failed
2020-12-16T09:42:29.8715510Z Generated patch file:
2020-12-16T09:42:29.8751911Z error: unrecognized input
2020-12-16T09:42:29.9423928Z ##[section]Finishing: Bash
```

**Comment by gleon99** [comment]
bot:pipe:retest

File: src/ucs/config/types.h (hunk 5 above)
Comment (yosefe) [reply src/ucs/config/types.h]: let's use regular enum values, not bits

Comment (gleon99) [reply src/ucs/config/types.h]: Done

**Comment by gleon99** [comment]
bot:pipe:retest

File: src/ucs/config/parser.c [hunk 7]
```diff
@@ -855,6 +858,84 @@ void ucs_config_help_array(char *buf, size_t max, const void *arg)
     array->parser.help(buf + strlen(buf), max - strlen(buf), array->parser.arg);
 }
 
+int ucs_config_sscanf_allow_list(const char *buf, void *dest, const void *arg)
+{
+    ucs_config_allow_list_t *field  = dest;
+    unsigned offset                 = 0;
+
+    if (buf != NULL && buf[0] == ucs_config_parser_negate) {
```
Comment (yosefe) [inline src/ucs/config/parser.c:866]: 1. pls add ( ) around != and == expressions
2. we can assume buf!=NULL, linke sscanf() does for example

File: src/ucs/config/parser.c [hunk 8]
```diff
@@ -855,6 +858,84 @@ void ucs_config_help_array(char *buf, size_t max, const void *arg)
...
+        field->mode = UCS_CONFIG_ALLOW_LIST_ALLOW;
+    }
+
+    if (buf == NULL || !ucs_config_sscanf_array(&buf[offset], dest, arg)) {
+        return 0;
+    }
+
+    if (field->array.count >= 1 && !strcmp(field->array.names[0], UCS_CONFIG_PARSER_ALL)) {
+        field->mode = UCS_CONFIG_ALLOW_LIST_ALLOW_ALL;
+        ucs_assert(field->array.count == 1);
```
Comment (yosefe) [inline src/ucs/config/parser.c:879]: maybe remove this assert?

File: src/ucs/config/parser.c [hunk 9]
```diff
@@ -855,6 +858,84 @@ void ucs_config_help_array(char *buf, size_t max, const void *arg)
...
+        offset++;
+    } else {
+        field->mode = UCS_CONFIG_ALLOW_LIST_ALLOW;
+    }
+
+    if (buf == NULL || !ucs_config_sscanf_array(&buf[offset], dest, arg)) {
+        return 0;
+    }
+
+    if (field->array.count >= 1 && !strcmp(field->array.names[0], UCS_CONFIG_PARSER_ALL)) {
```
Comment (yosefe) [inline src/ucs/config/parser.c:877]: `(field->array.count >= 1) && ..`

File: src/ucs/config/parser.c [hunk 10]
```diff
@@ -855,6 +858,84 @@ void ucs_config_help_array(char *buf, size_t max, const void *arg)
...
+    unsigned offset                 = 0;
+
+    if (buf != NULL && buf[0] == ucs_config_parser_negate) {
+        field->mode = UCS_CONFIG_ALLOW_LIST_NEGATE;
+        offset++;
+    } else {
+        field->mode = UCS_CONFIG_ALLOW_LIST_ALLOW;
+    }
+
+    if (buf == NULL || !ucs_config_sscanf_array(&buf[offset], dest, arg)) {
```
Comment (yosefe) [inline src/ucs/config/parser.c:873]: `&dest->array`

File: src/ucs/config/parser.c [hunk 11]
```diff
@@ -855,6 +858,84 @@ void ucs_config_help_array(char *buf, size_t max, const void *arg)
...
+int ucs_config_sprintf_allow_list(char *buf, size_t max, const void *src,
+                                  const void *arg)
+{
+    const ucs_config_allow_list_t *allow_list = src;
+    size_t offset                             = 0;
+
+    if (allow_list->mode == UCS_CONFIG_ALLOW_LIST_ALLOW_ALL) {
+        snprintf(buf, max, UCS_CONFIG_PARSER_ALL);
+        return 1;
+    } else if (allow_list->mode == UCS_CONFIG_ALLOW_LIST_NEGATE) {
```
Comment (yosefe) [inline src/ucs/config/parser.c:896]: can remove else

File: src/ucs/config/parser.c [hunk 12]
```diff
@@ -855,6 +858,84 @@ void ucs_config_help_array(char *buf, size_t max, const void *arg)
...
+}
+
+ucs_status_t ucs_config_clone_allow_list(const void *src, void *dest, const void *arg)
+{
+    const ucs_config_allow_list_t *src_list = src;
+    ucs_config_allow_list_t *dest_list      = dest;
+
+    dest_list->mode = src_list->mode;
+    if (src_list->array.count == 0) {
+        dest_list->array.count = 0;
```
Comment (yosefe) [inline src/ucs/config/parser.c:911]: why need special case?

File: src/ucs/config/parser.c [hunk 13]
```diff
@@ -855,6 +858,84 @@ void ucs_config_help_array(char *buf, size_t max, const void *arg)
...
+    ucs_config_release_array(&allow_list->array, arg);
+}
+
+void ucs_config_help_allow_list(char *buf, size_t max, const void *arg)
+{
+    const ucs_config_array_t *array = arg;
+
+    snprintf(
+        buf,
+        max, "comma-separated list (use \"all\" for including all items or \'^\' for negation) of: ");
```
Comment (yosefe) [inline src/ucs/config/parser.c:935]: IMO better break the string

File: src/ucs/config/parser.c (hunk 7 above)
Comment (gleon99) [reply src/ucs/config/parser.c:866]: 2 - If I recall correctly, not having this check failed the pipeline. Will remove it as retest.

File: src/ucs/config/parser.c (hunk 8 above)
Comment (gleon99) [reply src/ucs/config/parser.c:879]: Why? Isn't `all,some1,some2` an invalid input?

Comment (yosefe) [reply src/ucs/config/parser.c:879]: if it's invalid input we should return error rather than assert + gtest to check it returns err

File: src/ucs/config/parser.c [hunk 14]
```diff
@@ -863,21 +863,24 @@ int ucs_config_sscanf_allow_list(const char *buf, void *dest, const void *arg)
...
-    if (buf != NULL && buf[0] == ucs_config_parser_negate) {
+    if (buf[0] == ucs_config_parser_negate) {
         field->mode = UCS_CONFIG_ALLOW_LIST_NEGATE;
         offset++;
     } else {
         field->mode = UCS_CONFIG_ALLOW_LIST_ALLOW;
     }
 
-    if (buf == NULL || !ucs_config_sscanf_array(&buf[offset], dest, arg)) {
+    if (buf == NULL || !ucs_config_sscanf_array(&buf[offset], &field->array, arg)) {
```
Comment (yosefe) [inline src/ucs/config/parser.c:873]: need to remove buff==NULL here as well, should fix Coverity

File: src/ucs/config/parser.c [hunk 15]
```diff
@@ -863,21 +863,24 @@ int ucs_config_sscanf_allow_list(const char *buf, void *dest, const void *arg)
...
         return 0;
     }
 
-    if (field->array.count >= 1 && !strcmp(field->array.names[0], UCS_CONFIG_PARSER_ALL)) {
+    if ((field->array.count >= 1) &&
+        !strcmp(field->array.names[0], UCS_CONFIG_PARSER_ALL)) {
         field->mode = UCS_CONFIG_ALLOW_LIST_ALLOW_ALL;
-        ucs_assert(field->array.count == 1);
         ucs_config_release_array(&field->array, arg);
+        if(field->array.count != 1) {
```
Comment (yosefe) [inline src/ucs/config/parser.c:881]: space after if

File: src/ucs/config/parser.c [hunk 16]
```diff
@@ -863,21 +863,24 @@ int ucs_config_sscanf_allow_list(const char *buf, void *dest, const void *arg)
...
+            return 0;
+        }
```
Comment (yosefe) [inline src/ucs/config/parser.c:883]: pls add a space line after

File: test/gtest/ucs/test_config.cc [hunk 17]
```diff
@@ -623,3 +623,11 @@ UCS_TEST_F(test_config, test_allow_list) {
...
     }
 }
+
+UCS_TEST_F(test_config, test_allow_list_negative)
+{
+    ucs_config_allow_list_t field;
+
+    EXPECT_EQ(ucs_config_sscanf_allow_list("all,all", &field,
+                                           &ucs_config_array_string), 0);
+}
```
Comment (yosefe) [inline test/gtest/ucs/test_config.cc:633]: pls add a newline at the end of the file

File: src/ucs/config/parser.c [hunk 18]
```diff
@@ -920,10 +920,6 @@ void ucs_config_release_allow_list(void *ptr, const void *arg)
 {
     ucs_config_allow_list_t *allow_list = ptr;
 
-    if (allow_list->mode == UCS_CONFIG_ALLOW_LIST_ALLOW_ALL) {
```
Comment (yosefe) [inline src/ucs/config/parser.c]: aren't we already releasing the array in line 880?

Comment (gleon99) [reply src/ucs/config/parser.c]: Yes we do. The comment is on outdated diff, I removed it in `280867d`.

Comment (yosefe) [reply src/ucs/config/parser.c]: yes, a saw you removed. but seems like we should keep this 'if', otherwise it's double-free

Comment (gleon99) [reply src/ucs/config/parser.c]: Why double free? AFAIU, the `free()` itself happens only once, deep in `ucs_config_release_array`.

Comment (yosefe) [reply src/ucs/config/parser.c]: 1. call ucs_config_release_array after detecting that the value is "all", in ucs_config_sscanf_allow_list
2. call ucs_config_release_array from ucs_config_release_allow_list

**Review by yosefe — APPROVED** [review]

File: src/ucs/config/parser.c [hunk 19]
```diff
@@ -817,10 +817,12 @@ ucs_status_t ucs_config_clone_array(const void *src, void *dest, const void *arg
     ucs_status_t status;
     unsigned i;
 
-    dest_array->data = ucs_calloc(src_array->count, array->elem_size,
-                                  "config array");
-    if (dest_array->data == NULL) {
-        return UCS_ERR_NO_MEMORY;
+    if (src_array->count) {
+        dest_array->data = ucs_calloc(src_array->count, array->elem_size,
```
Comment (yosefe) [inline src/ucs/config/parser.c:821]: can dest_array->data be left un-initialized?

File: src/ucs/config/parser.c [hunk 20]
```diff
@@ -817,10 +817,12 @@ ucs_status_t ucs_config_clone_array(const void *src, void *dest, const void *arg
...
+                                    "config array");
```
Comment (yosefe) [inline src/ucs/config/parser.c:822]: indent

File: src/ucs/config/parser.c (hunk 19 above)
Comment (gleon99) [reply src/ucs/config/parser.c:821]: If it's `src_array->count`, there is no point in initializing  `dest_array->data` - anyway `data` is not used, and its release is skipped (actually valgrind even complains on that, if doing `calloc()` wuth 0, then not `free()`ing it).

Comment (yosefe) [reply src/ucs/config/parser.c:821]: 1. maybe set it to NULL to avoid garbage?
2. `if (src_array->count > 0) {`

Comment (gleon99) [reply src/ucs/config/parser.c]: Done

File: src/ucp/core/ucp_context.c [hunk 21]
```diff
@@ -800,7 +800,7 @@ static ucs_status_t ucp_check_tl_names(ucp_context_t *context)
     return UCS_OK;
 }
 
-const char* ucp_tl_bitmap_str(ucp_context_h context, uint64_t tl_bitmap,
+const char* ucp_tl_bitmap_str(ucp_context_h context, ucp_tl_bitmap_t *tl_bitmap,
```
Comment (yosefe) [inline src/ucp/core/ucp_context.c:803]: seems unrelated..

Comment (gleon99) [reply src/ucp/core/ucp_context.c]: Fixed

**Review by yosefe — APPROVED** [review]

**Comment by gleon99** [comment]
@yosefe  - squashed.

**Comment by gleon99** [comment]
bot:pipe:retest
